### DISCUSSION TITLE
Please provide documentation for deckgl charts

### DISCUSSION THREAD

**Discussion by songololo** [apache/superset#18417] [Ideas]
It is difficult to find any meaningful information about the deckgl charts.

Would it please be possible to add at least rudimentary documentation providing a brief overview of the accepted geospatial formats (e.g. GeoJSON) and any requirements such as coordinate reference system (e.g. epsg 4326?), geometry types (e.g. do Multipolygons work or only Polygons?) etc.

Thanks.

[a comment by an automated account is not shown]

**Comment by junlincc** [comment]
@srinify

**Comment by mistercrunch** [comment]
There's a bit of info here https://docs.preset.io/docs/scatterplot

**Reply by yagamiar15** [reply]
Page not found. Right now I'm facing some of the issues presented here. Currently working on a solution but I'm getting short of ideas

**Reply by jo-chemla** [reply]
> Page not found. 

Correct link is https://docs.preset.io/docs/time-series-scatter-plot

**Comment by songololo** [comment]
Thank you. Can anyone give some clues as to what the polygon chart expects in terms of geometry? (e.g. whether geojson polygon in epsg 4326 coordinate reference system works?)

**Comment by songololo** [comment]
@srinify @mistercrunch @junlincc

I'm able to get it working for point data (e.g. GeoJSON in EPSG 4326 CRS) but not for Polygon data:

Here are some examples that should be fairly easy to reproduce. These make use of the SQL editor and I'm using the amancevice docker container (v0.37.2).

Running a query to create a rectangular polygon (using PostGIS):
`SELECT 1 as mock_data, ST_MakeEnvelope(-0.127677, 51.507515, -0.122161, 51.510021, 4326) as geom;`

Gives a jagged polygon (instead of rectangular):
<img width="1585" alt="Screen Shot 2020-11-10 at 11 40 31" src="https://user-images.githubusercontent.com/4347878/98671432-26c71500-234c-11eb-946c-ee6811b121ed.png">

Similarly, when casting the geometry to Well Known Text format:
`SELECT 1 as mock_data, ST_AsText(ST_MakeEnvelope(-0.127677, 51.507515, -0.122161, 51.510021, 4326)) as geom;`

<img width="1589" alt="Screen Shot 2020-11-10 at 11 46 36" src="https://user-images.githubusercontent.com/4347878/98671479-3cd4d580-234c-11eb-82d8-21590a117656.png">

And with GeoJSON it displays nothing:
`SELECT 1 as mock_data, ST_AsGeoJSON(ST_MakeEnvelope(-0.127677, 51.507515, -0.122161, 51.510021, 4326)) as geom;`
<img width="1585" alt="Screen Shot 2020-11-10 at 11 49 50" src="https://user-images.githubusercontent.com/4347878/98671551-5aa23a80-234c-11eb-838c-41b7420c4c04.png">

Any information or advice on how to display a polygon would be much appreciated.

Thanks.

**Comment by songololo** [comment]
I managed to make some headway, posting here for others:

The JSON strings returned from `postGIS` using the `ST_AsGeoJSON` function contain only the geometry component of the geoJSON spec. This means that it is necessary to wrap these geometries in a full-fledged geoJSON object before `deck.gl` can recognise it.

For example:
```postgresql
SELECT *, json_build_object(
  'type',       'Polygon',
  'geometry',   ST_AsGeoJSON(ST_Transform(geom, 4326))::json)::text as geoJSON
FROM <my-schema>.<my-table>;
```

If you are working with MultiPolygons then you need to extract a Polygon first:
```postgresql
SELECT *, json_build_object(
  'type',       'Polygon',
  'geometry',   ST_AsGeoJSON(ST_Transform((ST_DUMP(geom)).geom::geometry(Polygon, 27700), 4326))::json)::text as geoJSON
FROM <my-schema>.<my-table>;
```

Note that in both cases, the geoJSON result has to be cast to a text type otherwise superset will have issues when it runs internal queries, which are otherwise unable to sort the results (duplicate checking?) if using JSON.

**Reply by yagamiar15** [reply]
First of all, thanks a lot for sharing insight in this kind of situation where there is a severe lack of documentation, this is pretty frustrating for me as I'm a jr Front-End dev. I'm tailoring the second function you provided as I'm working with multipolygons right now (geometry(multipolygon, 4326)) to be precise. 

``` 

SET search_path = <my-schema>, public;
SELECT geom , json_build_object(
  'type',       'Polygon',
  'geometry',   ST_AsGeoJSON(ST_Transform((ST_DUMP(geom)).geom::geometry(Polygon, 4326), 4326))::json)::text as geoJSON
FROM <my-schema>."<my-table>" 

```

Then I created a new table to copy the data and to the column that held the new transformed data I altered it as it follows

```

ALTER TABLE <my-schema>.<my-table>
ALTER COLUMN <my-column> TYPE text USING geojson::text

``` 

Then it finally started to work and show the polygons as intended.

Once again, thanks for your help! Otherwise I would have been totally lost

**Comment by mistercrunch** [comment]
Yes, I can confirm that we're expecting a geo json object as text. We could do more magic to understand the Postgres/postGIS geoJSON type and convert it accordingly, but it gets tricky for us to support special types that are database specific.

**Comment by sazary** [comment]
@songololo I can't tell you how much you've helped me with this! i have been trying for some time, even asked some questions here and in stackoverflow, and no answer... until I saw you solution, and it worked! Thanks dude

**Comment by flat-linex** [comment]
I have the same problem (I think) but I do not understand the solution.
I create the input for deck.gl json in superset using the attached sql
which results in the attached 'png' data.
Whene using the geojson field I get nothing from deck.gl within superset, when I use geojson_build I get an error:

> An error occurred while rendering the visualization: TypeError: can't convert undefined to object

Any help would be appreciated, thx


[postgis.txt](https://github.com/apache/superset/files/8080826/postgis.txt)
![postgis_snap](https://user-images.githubusercontent.com/11737850/154292180-4dc7e458-89a8-4042-8946-4188d6656382.png)

**Reply by songololo** [reply]
@flat-linex are you able to share more information about your queries and query results?

**Comment by juni3227** [comment]
I am having the same problem. This very fine geojson won't display any polygons.

```
{
    "type": "MultiPolygon",
    "coordinates": [
        [
            [
                [
                    128.68062301861633,
                    35.78292571606495
                ],
                [
                    128.68063634982013,
                    35.78382715110716
                ],
                [
                    128.68174263970826,
                    35.783816283155566
                ],
                [
                    128.68172929601977,
                    35.782914848471634
                ],
                [
                    128.68062301861633,
                    35.78292571606495
                ]
            ]
        ]
    ]
}
```

**Reply by juni3227** [reply]
What do you mean by fully fledged geojson? this is per spec, as far as I know.
The example over the website just ... has coordinates. 
supplying superset with just this will work, which is ... not a geojson.
```
            [
                [
                    128.68062301861633,
                    35.78292571606495
                ],
                [
                    128.68063634982013,
                    35.78382715110716
                ],
                [
                    128.68174263970826,
                    35.783816283155566
                ],
                [
                    128.68172929601977,
                    35.782914848471634
                ],
                [
                    128.68062301861633,
                    35.78292571606495
                ]
            ]
```
